File: Lab-12/examples/tcp_multithreaded_server.cpp
#include <iostream>
#include <string>
#include <ctime>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <netdb.h>
#include <pthread.h>

#define BUFFER_SIZE 1024

/**
g++ tcp_multithreaded_server.cpp -o tcp_multithreaded_server -lpthread
*/

void *ClientThread(void *args)
{
    int sock = *(int*)args;

    // odbieranie/wysylanie wiadomosci od/do klienta
    // ...
}

int main(int argc, char **argv)
{
    // utworzenie gniazda
    // bind
    // listen

    // akceptowanie klientow - w petli nieskonczonej
    while(client_fd = accept(server_fd, (struct sockaddr *) &client, &client_len))
    {
        // tworzymy nowy watek
        pthread_t thread;

        // uruchamiamy watek, jego obsluga (tym, za co odpowiada watek) zajmuje
        // sie funkcja ClientThread, przekazujemy jej jako argument socket klienta
        if(pthread_create(&thread, NULL, ClientThread, (void*) &client_fd) < 0)
        {
            std::cout << ("Could not create a new thread socket\n");
            perror("pthread_create");
            return 1;
        }
    }

    // zamkniecie socketow

    return 0;
}

File: Lab-12/examples/tcp_multithreaded_server.java
class ClientThread extends Thread {

    private Socket clientSocket;

    public ClientThread(Socket client) {
        this.clientSocket = client;
    }

    @Override
    public void run() {

        try {

            // wysylanie/odbieranie danych do/od klienta

            clientSocket.close();

        } catch (IOException ex) { }
    }
}

class Server {

    private Socket client_connection = null;
    private ServerSocket server = null;
    private int port = 0;

    public Server(int port) {
        this.port = port;
    }

    public void start() {
        try {

            this.server = new ServerSocket(this.port);

            while (true) {
                
                client_connection = server.accept();

                ClientThread c = new ClientThread(client_connection);
                	
                c.start();
            }

        } catch (IOException ex) { }
    }
}

public class tcp_multithreaded_server {

    public static void main(String[] args) {
        
        Server s = new Server(6666);
        s.start();
    }  
} 
